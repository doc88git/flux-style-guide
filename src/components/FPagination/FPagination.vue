<template>
  <section class="f-pagination">
    <ul>
      <li>
        <button :disabled="isFirstPage" @click="jumpTo('first')">
          Primeira
        </button>
      </li>
      <li>
        <button :disabled="isFirstPage" @click="jumpTo('prev')">
          <f-icon name="chevron-left" />
        </button>
      </li>
      <li v-for="i in show" :key="i">
        <button :class="getPageClasses(i)" @click="setCurrentPage(i)">
          {{ i }}
        </button>
      </li>
      <li>
        <button :disabled="isLastPage" @click="jumpTo('next')">
          <f-icon name="chevron-right" />
        </button>
      </li>
      <li>
        <button :disabled="isLastPage" @click="jumpTo('last')">
          Ãšltima
        </button>
      </li>
    </ul>
  </section>
</template>

<script>
import { FIcon } from '../FIcon'

export default {
  name: 'f-pagination',
  components: {
    FIcon
  },

  props: {
    currentPage: {
      type: Number,
      required: true
    },
    total: {
      type: Number,
      required: true
    },
    perPage: {
      type: Number,
      required: true
    },
    max: {
      type: Number,
      default: 10
    }
  },

  data: () => ({
    localCurrentPage: 0
  }),

  computed: {
    isFirstPage() {
      return this.localCurrentPage === 1
    },

    isLastPage() {
      return this.localCurrentPage >= this.totalPages
    },

    totalPages() {
      if (!this.total || !this.perPage) return 0
      return Math.ceil(this.total / this.perPage)
    },

    pgFrom() {
      return this.localCurrentPage - Math.ceil(this.max / 2)
    },

    show() {
      const result = Array.from({ length: this.max }, (e, i) =>
        this.pgFrom + this.max > this.totalPages
          ? this.totalPages + 1 - (this.max - i)
          : this.pgFrom <= 0
          ? i + 1
          : this.pgFrom + i
      )

      return result
    }
  },

  methods: {
    getPageClasses(i) {
      return { selected: this.localCurrentPage === i }
    },

    setCurrentPage(value) {
      this.localCurrentPage = value
      this.$emit('update:current_page', value)
    },

    jumpTo(position) {
      const value = parseInt(this.localCurrentPage)

      if (position === 'first') this.setCurrentPage(1)
      if (position === 'last') this.setCurrentPage(this.totalPages)

      if (position === 'prev' && value > 0) this.setCurrentPage(value - 1)

      if (position === 'next' && value <= this.totalPages)
        this.setCurrentPage(value + 1)
    },

    initPagination() {
      this.localCurrentPage = parseInt(this.currentPage)
      this.lastPage = this.totalPages.length - 1
    }
  },

  created() {
    this.initPagination()
  }
}
</script>

<style lang="scss" scoped>
.f-pagination {
  user-select: none;

  .f-icon {
    transform: translateY(2px);
    min-width: 35px;
    fill: var(--color-gray);

    &.selected {
      fill: var(--color-primary);
    }

    &:hover {
      fill: var(--color-primary-light);
    }

    &:disabled {
      opacity: 50%;
      cursor: default;
    }
  }

  ul,
  li {
    display: inline-block;
    min-height: 100%;
    text-align: center;
  }

  ul {
    display: flex;
    align-items: center;
    align-content: center;
    list-style-type: none;

    li {
      button {
        text-transform: capitalize;
        padding: 0;
        text-align: center;
        min-width: 35px;
        outline: none;
        color: var(--color-gray);

        &.selected {
          color: var(--color-primary);
        }

        &:hover {
          color: var(--color-primary-light);
        }

        &:disabled {
          opacity: 50%;
          cursor: default;
        }
      }
    }
  }
}
</style>
