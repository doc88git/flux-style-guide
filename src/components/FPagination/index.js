import FPagination from "./FPagination";

export { FPagination };
