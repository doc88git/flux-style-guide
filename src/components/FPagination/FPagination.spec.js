import { shallowMount } from '@vue/test-utils'
import FPagination from './FPagination'

const WRAPPER_PROPS = { currentPage: 1, total: 100, perPage: 6, max: 10 }


describe('FPaginantion tests', () => {
  let WRAPPER

  beforeEach(() => WRAPPER = shallowMount(FPagination, { propsData: WRAPPER_PROPS }))

  test('Componente existe', () => {
    expect(WRAPPER.exists()).toBe(true)
  })

  describe('Teste de computeds', () => {

    describe('Computed totalPages', () => {
      test('totalPages quando a prop total: 100 e perPage: 6', () => {
        expect(WRAPPER.vm.totalPages).toBe(17)
      })

      test('totalPages quando a prop total: 200 e perPage: 6', () => {
        WRAPPER.setProps({ total: 200 })
        expect(WRAPPER.vm.totalPages).toBe(34)
      })

      test('totalPages quando a prop total: 50 e perPage: 2', () => {
        WRAPPER.setProps({ total: 50, perPage: 2 })
        expect(WRAPPER.vm.totalPages).toBe(25)
      })
    })

    describe('Computed show', () => {
      test('show quando localCurrentPage: 1', () => {
        expect(WRAPPER.vm.show).toStrictEqual([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
      })

      test('show quando localCurrentPage: 11', () => {
        WRAPPER.setData({ localCurrentPage: 11 })
        expect(WRAPPER.vm.show).toStrictEqual([6, 7, 8, 9, 10, 11, 12, 13, 14, 15])
      })

      test('show quando localCurrentPage: 15', () => {
        WRAPPER.setData({ localCurrentPage: 15 })
        expect(WRAPPER.vm.show).toStrictEqual([8, 9, 10, 11, 12, 13, 14, 15, 16, 17])
      })

      test('show quando localCurrentPage: 18', () => {
        WRAPPER.setData({ localCurrentPage: 18 })
        expect(WRAPPER.vm.show).toStrictEqual([8, 9, 10, 11, 12, 13, 14, 15, 16, 17])
      })
    })

    describe('Computed isFirstPage', () => {
      test('isFirstPage quando localCurrentPage: 1', () => {
        expect(WRAPPER.vm.isFirstPage).toBeTruthy()
      })

      test('isFirstPage quando localCurrentPage: 10', () => {
        WRAPPER.setData({ localCurrentPage: 10 })
        expect(WRAPPER.vm.isFirstPage).toBeFalsy()
      })
    })

    describe('Computed isLastPage', () => {
      test('isLastPage quando localCurrentPage: 1', () => {
        expect(WRAPPER.vm.isLastPage).toBeFalsy()
      })

      test('isLastPage quando localCurrentPage: 17', () => {
        WRAPPER.setData({ localCurrentPage: 17 })
        expect(WRAPPER.vm.isLastPage).toBeTruthy()
      })
    })
  })

  describe('Teste de métodos', () => {

    describe('Teste do método jumpTo', () => {
      test('Pular para próxima página', () => {
        WRAPPER.vm.jumpTo('next')
        expect(WRAPPER.vm.localCurrentPage).toBe(2)
      })

      test('Pular para página anterior', () => {
        WRAPPER.setData({ localCurrentPage: 10 })
        WRAPPER.vm.jumpTo('prev')
        expect(WRAPPER.vm.localCurrentPage).toBe(9)
      })

      test('Pular para ultima página', () => {
        WRAPPER.vm.jumpTo('last')
        expect(WRAPPER.vm.localCurrentPage).toBe(17)
      })

      test('Pular para primeira página', () => {
        WRAPPER.vm.jumpTo('first')
        expect(WRAPPER.vm.localCurrentPage).toBe(1)
      })
    })

    describe('Teste do método setCurrentPage', () => {

      test('setCurrentPage com argumento 10 e emissão do evento update:current_page', async () => {
        WRAPPER.vm.setCurrentPage(10)
        expect(WRAPPER.vm.localCurrentPage).toBe(10)

        // Wait util $emit has been handled and check for
        // event emission.
        await WRAPPER.vm.$nextTick()

        // The emitted() function returns the event as an Array
        // within an Array, for each event emmited.
        expect(WRAPPER.emitted()['update:current_page'][0][0]).toBe(10)
      })

      test('setCurrentPage com argumento 0 e emissão do evento update:current_page', async () => {
        WRAPPER.vm.setCurrentPage(0)
        expect(WRAPPER.vm.localCurrentPage).toBe(0)

        await WRAPPER.vm.$nextTick()
        expect(WRAPPER.emitted()['update:current_page'][0][0]).toBe(0)
      })
    })
  })

  test('Componente não alterado', () => {
    expect(WRAPPER.html()).toMatchSnapshot()
  })
})