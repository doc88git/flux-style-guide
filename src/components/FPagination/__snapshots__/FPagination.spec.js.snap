// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FPaginantion tests Componente não alterado 1`] = `
"<section class=\\"FPagination\\">
  <div class=\\"FPagination__sectionPrevious\\"><button disabled=\\"disabled\\" class=\\"FPagination__sectionBtn\\">
      Primeira
    </button> <button disabled=\\"disabled\\" class=\\"FPagination__sectionBtn\\">
      <f-icon-stub color=\\"gray-300\\" lib=\\"flux\\" size=\\"base\\" name=\\"chevron-left\\"></f-icon-stub>
    </button></div>
  <ul class=\\"FPagination__ul\\">
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn FPagination__pageBtn--selected\\">
        1
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        2
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        3
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        4
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        5
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        6
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        7
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        8
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        9
      </button></li>
    <li class=\\"FPagination__li\\"><button class=\\"FPagination__pageBtn\\">
        10
      </button></li>
  </ul>
  <div class=\\"FPagination__sectionNext\\"><button class=\\"FPagination__sectionBtn\\">
      <f-icon-stub color=\\"gray\\" lib=\\"flux\\" size=\\"base\\" name=\\"chevron-right\\"></f-icon-stub>
    </button> <button class=\\"FPagination__sectionBtn\\">
      Última
    </button></div>
</section>"
`;
